File: app/src/main/kotlin/metamark/parser/MatchOneOrMoreExpressionTerm.kt
package metamark.parser

import metamark.parser.antlr.MetamarkParser

data class MatchOneOrMoreExpressionTerm(
    val repeatedExpression: ExpressionTerm,
): ExpressionTerm() {
    companion object {
        fun build(
            ctx: MetamarkParser.Expression_matchOneOrMoreExpressionContext,
        ): MatchOneOrMoreExpressionTerm {
            return MatchOneOrMoreExpressionTerm(
                repeatedExpression = ExpressionTerm.build(ctx.repeatedExpression)
            )
        }
    }
}

File: MetaMark/src/metamark_ast/MatchOneOrMoreExpressionTerm.ts
import { ExpressionTerm } from "./ExpressionTerm";
import { Expression_matchOneOrMoreExpressionContext } from "../../generated_src/MetamarkParser";
import { ExpressionTermUtils } from "./ExpressionTermUtils";
import { DelegatingExpressionTerm } from "./DelegatingExpressionTerm";
import { StaticScope } from "./StaticScope";

export class MatchOneOrMoreExpressionTerm extends DelegatingExpressionTerm {
    static equals(
        a: MatchOneOrMoreExpressionTerm,
        b: MatchOneOrMoreExpressionTerm,
    ): boolean;
    static equals(a: unknown, b: unknown): boolean | undefined;

    static equals(a: unknown, b: unknown): boolean | undefined {
        if (!(a instanceof MatchOneOrMoreExpressionTerm)) return undefined;
        if (!(b instanceof MatchOneOrMoreExpressionTerm)) return undefined;
        return ExpressionTermUtils.equals(
            a.repeatedExpression,
            b.repeatedExpression,
        );
    }

    static build(
        staticScope: StaticScope,
        ctx: Expression_matchOneOrMoreExpressionContext,
    ): MatchOneOrMoreExpressionTerm {
        return new MatchOneOrMoreExpressionTerm(
            ExpressionTermUtils.build(staticScope, ctx._repeatedExpression),
        );
    }

    readonly repeatedExpression: ExpressionTerm;

    constructor(repeatedExpression: ExpressionTerm) {
        super("parseOneOrMore", repeatedExpression);

        this.repeatedExpression = repeatedExpression;
    }
}
